### PR TITLE
Support Alignment Extraction for all RNNT Beam decoding methods

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Adds alignment extraction for all rnnt beam algorithms

**Collection**: [ASR]

# Changelog 
- Adds support for TSD, ALSD and mAES alignment extraction 
- Adds unittest that downloads a small model and runs on AN4 test set to compute alignments from greedy and beam search algorithms

# Usage
You can now submit `preserve_alignments=true` to in the RNNTDecodingConfig for all beam search algos and extract alignments.

```python

import nemo.collections.asr as nemo_asr

filepath = "<audio filepath>"

model = nemo_asr.models.ASRModel.from_pretrained('stt_en_conformer_transducer_large')  # type: nemo_asr.models.EncDecRNNTBPEModel
# Create default decoding config
model.change_decoding_strategy(None)
decoding_cfg = model.cfg.decoding
decoding_cfg.strategy = "beam"  # can be beam, tsd, alsd or maes
decoding_cfg.preserve_alignments = True  # the important argument !
decoding_cfg.beam.beam_size = 8
decoding_cfg.beam.return_best_hypothesis = False

# Prepare beam search decoding strategy
model.change_decoding_strategy(decoding_cfg)

best_hypothesis, all_hypothesis = model.transcribe([filepath], return_hypotheses=True)
print("Num hypothesis per sample :", len(best_hypothesis[0]))
print("Result type :", best_hypothesis[0][0].__class__.__name__)
for candidate in range(len(best_hypothesis[0])):
    print(f"Beam Search {candidate + 1} :", best_hypothesis[0][candidate].text, best_hypothesis[0][candidate].score)

# Use the following commented print statements to check
# the alignment of other algorithms compared to the default
all_hyps = all_hypothesis
for idx, hyp_ in enumerate(all_hyps):  # type: (int, rnnt_utils.Hypothesis)
    print("Hyp index", idx + 1, "text :", hyp_.text)

    for t in range(len(hyp_.alignments)):
        t_u = []
        for u in range(len(hyp_.alignments[t])):
            logp, label = hyp_.alignments[t][u]
            t_u.append(int(label))
        print(f"Tokens at timestep {t} = {t_u}")
    print()

```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: tests/collections/asr/decoding/test_rnnt_decoding.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        with torch.no_grad():
+            hyps = beam(encoder_output=enc_out, encoded_lengths=enc_len)[0]  # type: rnnt_utils.Hypothesis
+            hyp = decode_text_from_greedy_hypotheses(hyps, model.decoding)
+            hyp = hyp[0]
+
+            assert hyp.alignments is not None
+
+            # Use the following commented print statements to check
+            # the alignment of other algorithms compared to the default
+            # print("Text", hyp.text)
```
Comment (Kipok) [inline tests/collections/asr/decoding/test_rnnt_decoding.py:170]: I usually prefer to leave the print statements "on" all the time - by default test output is not being printed, but if something fails, it's very helpful to see such messages right away without the need to change code and rerun

File: tests/collections/asr/decoding/test_rnnt_decoding.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                    for u in range(len(hyp_.alignments[t])):
+                        logp, label = hyp_.alignments[t][u]
+                        assert torch.is_tensor(logp)
+                        assert torch.is_tensor(label)
+
+                        t_u.append(int(label))
+
+                    # Blank token must be the last token in the current
+                    if len(t_u) > 1:
+                        assert t_u[-1] == blank_id
```
Comment (Kipok) [inline tests/collections/asr/decoding/test_rnnt_decoding.py:245]: Can we also add a check that all other tokens are non-blank? That was breaking in the bug-fix PR I did for batched alignment calculation, so might be a helpful check here as well

File: tests/collections/asr/decoding/test_rnnt_decoding.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+    @pytest.mark.skipif(
+        not NUMBA_RNNT_LOSS_AVAILABLE, reason='RNNTLoss has not been compiled with appropriate numba version.',
+    )
+    @pytest.mark.with_downloads
+    @pytest.mark.unit
+    @pytest.mark.parametrize(
+        "beam_config",
+        [
+            {"search_type": "greedy"},
```
Comment (Kipok) [inline tests/collections/asr/decoding/test_rnnt_decoding.py:191]: I don't really know the details about all these algorithms, but if any of them are expected to produce exactly the same label sequence, that would be a nice thing to test for as well. Or maybe something like greedy = beam with beam size of 1. Very different code paths but should give equivalent results - always a great test to have, in my opinion.

File: nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py [hunk 4]
```diff
@@ -425,7 +429,8 @@ def greedy_search(
             not_blank = True
             symbols_added = 0
 
-            while not_blank and (symbols_added < self.max_candidates):
+            # TODO: Figure out how to remove this hard coding afterwords
+            while not_blank and (symbols_added < 5):
```
Comment (Kipok) [inline nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py:433]: Why do we have 5 here?

**Review by Kipok — COMMENTED** [review]
Looks generally good to me with a few comments. I should say that I don't really know the details of all of those algorithms, so will be great if someone with more knowledge about them can also have a look before merging.

File: nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py:433]: Heuristic. Depending on model stride (T / U ratio ~ 2) so the model could at worst predict 1 token every 2 timesteps - we give it some leeway with up to 5 tokens per timestep. Normally it will never reach such a high number, even for 8x stride.

File: tests/collections/asr/decoding/test_rnnt_decoding.py (hunk 1 above)
Comment (titu1994) [reply tests/collections/asr/decoding/test_rnnt_decoding.py]: Agreed, made them print always

File: tests/collections/asr/decoding/test_rnnt_decoding.py (hunk 3 above)
Comment (titu1994) [reply tests/collections/asr/decoding/test_rnnt_decoding.py:189]: It auto switches to greedy decoding at beam size 1 (so there's no actual beam search being done with beam size 1)

File: tests/collections/asr/decoding/test_rnnt_decoding.py (hunk 2 above)
Comment (titu1994) [reply tests/collections/asr/decoding/test_rnnt_decoding.py:244]: Nice catch, will add.

File: tests/collections/asr/decoding/test_rnnt_decoding.py (hunk 3 above)
Comment (Kipok) [reply tests/collections/asr/decoding/test_rnnt_decoding.py:189]: Maybe we can add some flag to allow to keep beam search logic with beam size of 1? Useful for the tests + might be quite useful for debugging issues as well.

File: nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py (hunk 4 above)
Comment (Kipok) [reply nemo/collections/asr/parts/submodules/rnnt_beam_decoding.py:433]: Got it - maybe mention this in the comment, so that it's clear why we have this constant here

File: tests/collections/asr/decoding/test_rnnt_decoding.py (hunk 3 above)
Comment (titu1994) [reply tests/collections/asr/decoding/test_rnnt_decoding.py:189]: The beam search logic doesn't work for size 1, but I can look info it in a separate pr
